File: ts/samples/dapr-sample/packages/domain/src/aggregates/weatherForecast/projectors/weather-forecast-projector.ts
import { 
  defineProjector, 
  EmptyAggregatePayload, 
  AggregateProjector, 
  PartitionKeys, 
  Aggregate, 
  IEvent, 
  Result, 
  ok, 
  SekibanError 
} from '@sekiban/core';
import { z } from 'zod';
import { 
  WeatherForecastInputted, 
  WeatherForecastLocationUpdated, 
  WeatherForecastDeleted 
} from '../events/weather-forecast-events.js';
import { 
  WeatherForecastState, 
  DeletedWeatherForecastState, 
  WeatherForecastPayloadUnion 
} from '../payloads/weather-forecast-payloads.js';

export const weatherForecastProjectorDefinition = defineProjector<WeatherForecastPayloadUnion>({
  aggregateType: 'WeatherForecast',
  
  initialState: () => new EmptyAggregatePayload(),
  
  projections: {
    WeatherForecastInputted: (state: any, event: z.infer<typeof WeatherForecastInputted.schema>) => ({
      aggregateType: 'WeatherForecast' as const,
      location: event.location,
      date: event.date,
      temperatureC: event.temperatureC,
      summary: event.summary
    } as WeatherForecastState),
    
    WeatherForecastLocationUpdated: (state: any, event: z.infer<typeof WeatherForecastLocationUpdated.schema>) => {
      if (!state || state.aggregateType !== 'WeatherForecast') return state;
      return {
        ...state,
        location: event.location
      } as WeatherForecastState;
    },
    
    WeatherForecastDeleted: (state: any, event: z.infer<typeof WeatherForecastDeleted.schema>) => {
      if (!state || state.aggregateType !== 'WeatherForecast') return state;
      return {
        aggregateType: 'DeletedWeatherForecast' as const
      } as DeletedWeatherForecastState;
    }
  }
});

export class WeatherForecastProjector extends AggregateProjector<WeatherForecastPayloadUnion> {
  readonly aggregateTypeName = 'WeatherForecast';
  readonly multiProjectorName = 'WeatherForecastAggregateListProjector';
  
  getInitialState(partitionKeys: PartitionKeys): Aggregate<EmptyAggregatePayload> {
    return weatherForecastProjectorDefinition.getInitialState(partitionKeys);
  }
  
  project(
    aggregate: Aggregate<WeatherForecastPayloadUnion | EmptyAggregatePayload>, 
    event: IEvent
  ): Result<Aggregate<WeatherForecastPayloadUnion | EmptyAggregatePayload>, SekibanError> {
    return weatherForecastProjectorDefinition.project(aggregate, event);
  }
  
  canHandle(eventType: string): boolean {
    return [
      'WeatherForecastInputted',
      'WeatherForecastLocationUpdated',
      'WeatherForecastDeleted'
    ].includes(eventType);
  }
  
  getSupportedPayloadTypes(): string[] {
    return ['WeatherForecast', 'DeletedWeatherForecast'];
  }
}

File: pure/internalUsages/SharedDomain/Aggregates/WeatherForecasts/WeatherForecastProjector.cs
using Sekiban.Pure.Aggregates;
using Sekiban.Pure.Events;
using Sekiban.Pure.Projectors;
using SharedDomain.Aggregates.WeatherForecasts.Events;
using SharedDomain.Aggregates.WeatherForecasts.Payloads;
namespace SharedDomain.Aggregates.WeatherForecasts;

[GenerateSerializer]
public class WeatherForecastProjector : IAggregateProjector
{
    public Type[] PayloadTypes => new[]
    {
        typeof(WeatherForecast),
        typeof(DeletedWeatherForecast)
    };

    public IAggregatePayload Project(IAggregatePayload payload, IEvent ev) =>
        (payload, ev.GetPayload()) switch
        {
            (EmptyAggregatePayload, WeatherForecastInputted inputted) => new WeatherForecast
            {
                Location = inputted.Location,
                Date = inputted.Date,
                TemperatureC = inputted.TemperatureC,
                Summary = inputted.Summary
            },
            (WeatherForecast forecast, WeatherForecastLocationUpdated locationUpdated) => forecast with
            {
                Location = locationUpdated.Location
            },
            (WeatherForecast forecast, WeatherForecastDeleted) => DeletedWeatherForecast.Instance,
            _ => payload
        };
}
